text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,313 @@
+
+However, in practice, having a collator signature in the receipt on the relay chain does not 
+provide any benefits as there is no mechanism to punish or reward collators that have provided 
+bad parachain blocks.
+
+This proposal aims to remove the collator signature and all the logic that checks the collator 
+signatures of candidate receipts. We use the first 7 reclaimed bytes to represent version, 
+the core and session index, and fill the rest with zeroes. So, there is no change in the layout 
+and length of the receipt. The new primitive is binary compatible with the old one.
+
+### UMP transport
+
+[CandidateCommitments](https://github.com/paritytech/polkadot-sdk/blob/b5029eb4fd6c7ffd8164b2fe12b71bad0c59c9f2/polkadot/primitives/src/v7/mod.rs#L682) 
+remains unchanged as we will store scale encoded `UMPSignal` messages directly in the parachain 
+UMP queue by outputing them in [upward_messages](https://github.com/paritytech/polkadot-sdk/blob/b5029eb4fd6c7ffd8164b2fe12b71bad0c59c9f2/polkadot/primitives/src/v7/mod.rs#L684). 
+
+The UMP queue layout is changed to allow the relay chain to receive both the XCM messages and 
+`UMPSignal` messages. An empty message (empty `Vec<u8>`) is used to mark the end XCM messages and 
+the start of `UMPSignal` messages.
+
+This way of representing the new messages has been chosen over introducing an enum wrapper to 
+minimize breaking changes of XCM message decoding in tools like Subscan for example. 
+
+Example: 
+```rust
+[ XCM message1, XCM message2, ..., EMPTY message, UMPSignal::SelectCore ]
+```
+
+#### `UMPSignal` messages
+
+```rust
+/// An `u8` wrap around sequence number. Typically this would be the least significant byte of the 
+/// parachain block number.
+pub struct CoreSelector(pub u8);
+
+/// An offset in the relay chain claim queue.
+pub struct ClaimQueueOffset(pub u8);
+
+/// Signals sent by a parachain to the relay chain.
+pub enum UMPSignal {
+	/// A message sent by a parachain to select the core the candidate is committed to.
+	/// Relay chain validators, in particular backers, use the `CoreSelector` and `ClaimQueueOffset`
+	/// to compute the index of the core the candidate has committed to.
+	///
+	SelectCore(CoreSelector, ClaimQueueOffset),
+}
+```
+
+The parachain runtime is not concerned with the actual `CoreIndex` the candidate is intended for, 
+but must provide enough information for the validators and collators to compute it. `CoreSelector` 
+and `ClaimQueueOffset` use only 2 bytes and fulfill the requirement.
+
+**Example:**
+
+`cq_offset = 1` and `core_selector = 3`
+
+The table below represents a snapshot of the claim queue:
+
+|  | offset = 0 | offset = 1 | offset = 2 |
+| :--:	 | :--:   | :--:   | :--:   |
+| Core 1    | **Para A**     | **Para A**     | **Para A**     |
+| Core 2   | **Para A**  | Para B  | **Para A**  |
+| Core 3   | Para B  | **Para A**  | **Para A** |
+
+The purpose of `ClaimQueueOffset` is to select the column from the above table. 
```

Comment:
Does Polkadot right now allows to use a claim ahead in the queue? I still don't really get why we need it. The parachain runtime should ensure that always a new core is being used, for that using selector is enough.
Do we use right now the relay chain block of the pov was build on to determine the claim queue and then to determine the core? Or what are we doing right now?

Reply:
Right now, the core index is determined by the backers who have voted. We map validator index to group index to core index. 

The claim queue offset is required because of the `CoreSelector % number_of_assignments` operation that selects the core. Because parachains can have a varying number of cores assigned to them (on-demand) at a given RCB, the core index you get might be different depending on which claim is used.

Reply:
I mean I get this. However, you want to us all possible claims? So, you probably don't want to waste any cores. Which means you will build on all the cores. With prospective parachains you also can calculate which should be the next claim to use. Thus, we should not require the offset?

Reply:
> I mean I get this. However, you want to us all possible claims? So, you probably don't want to waste any cores. Which means you will build on all the cores.

Yes

> With prospective parachains you also can calculate which should be the next claim to use. Thus, we should not require the offset?

That would not be enough. In backing we need to determine if the candidate is valid on the core the backers are assigned to. We take the core selector and determine whatever core comes out of that. However, if one of these blocks doesn't make it in the next relay chain block, the relay chain cannot validate the candidate core index anymore. At that point the top of the claim queue might have a different number of assignments and you'd get a different core index for the modulo operation.

Reply:
I thought about this a little bit more and I think I found a way to not require the `offset`. My main problem with this `offset` is that it is static. You also write in the RFC itself that it isn't that great. 

So, to my proposal. I would propose we store the latest `core_selector` of each parachain in the relay chain state. Then we can use this `latest_core_selector` (fetched at the candidate context relay chain block) together with the `claim_queue` to determine if the `core_index` in the candidate is correct. First we would calculate the `core_selector_diff = core_selector.wrapping_sub(latest_core_selector)`. Then we would fetch the `claim_queue` at the candidate context relay chain block. We would count all cores assigned to our parachain until the count reaches `core_selector_diff` and then we would have our `core_index`. 

I think we can then also get rid off `max_candidate_len` from the async backing parameters as the maximum length would be determined by the "claim queue lookahead".

Reply:
For `max_candidate_len` see: https://github.com/paritytech/polkadot-sdk/issues/5079 -> is already planned.

For the rest. I am not following: 

> We would count all cores assigned to our parachain

This is not deterministic, if you don't know the position the parachain is aiming for.

> My main problem with this offset is that it is static.

It does not have to be. It just will be for most parachains as I don't see much reason to change it all the time. But the value could be adjustable by parachain governance for example.

I would expect most parachains to use 1, some maybe 0 to get the lowest latency and then maybe some to use 2 (with an older relay parent) to get more stable/performant block production at the cost of latency when it comes to message processing.

Reply:
> It does not have to be. It just will be for most parachains as I don't see much reason to change it all the time. But the value could be adjustable by parachain governance for example.

IMO exposing this parameter to parachain builders will be more confusing to them. They already don't know what the unincluded segment is. The unincluded segment used by the lookahead collator would not work anymore with the static offset, because it builds multiple blocks on the same relay chain block.

> This is not deterministic, if you don't know the position the parachain is aiming for.

Why is this not deterministic? The claim queue doesn't change in the past and could only change for future blocks, because of some ondemand core for example. When the collator is sending the collation, it is aware of the claim queue and where it sends the PoV. 


I also think that we will have problems with on demand? Let's assume we are able to buy a core and assign it in the same block (I know there exists somewhere an issue for this). How should the collator be able to build a block on time? Maybe making the UMP optional could be a solution for this? 

So, maybe let me propose the following:

1. We re-arrange the claim queue storage, to be able to include the claim queue in the storage proof used by the parachain. This way the parachain can report the `core` directly it wants to be scheduled on. We don't need any offset nor any selector.

2. The UMP message is made optional. This way the runtime gives up the "control" to the PoV submitter to decide on the core. This gives us the freedom we have right now and will work for all chains that don't use elastic scaling without a problem.

Reply:
> > It does not have to be. It just will be for most parachains as I don't see much reason to change it all the time. But the value could be adjustable by parachain governance for example.
> 
> IMO exposing this parameter to parachain builders will be more confusing to them. They already don't know what the unincluded segment is. 

The way this is being implemented right now doesn't require the parachain builders to care about this parameter - for most usecases `1` is a good value.
> The unincluded segment used by the lookahead collator would not work anymore with the static offset, because it builds multiple blocks on the same relay chain block.
> 

I don't understand why it wouldn't work if it builds on the same relay parent. Do you mean it can't build on all of the cores assigned by claim queue ahead of time ?

> > This is not deterministic, if you don't know the position the parachain is aiming for.
> 
> Why is this not deterministic? The claim queue doesn't change in the past and could only change for future blocks, because of some ondemand core for example. When the collator is sending the collation, it is aware of the claim queue and where it sends the PoV.
> 
> I also think that we will have problems with on demand? Let's assume we are able to buy a core and assign it in the same block (I know there exists somewhere an issue for this). How should the collator be able to build a block on time? Maybe making the UMP optional could be a solution for this?
> 

I don't think that making UMP optional helps, because you might want to get `2` cores in same relay chain block and you need the UMP commitment. For the current proposal, you would want a CQ offset of `0`. However, it would likely be a good idea to always get on-demand cores assigned with at least 1 RCB delay to make sure there is plenty of time for the collation to be built ahead of time.

> So, maybe let me propose the following:
> 
> 1. We re-arrange the claim queue storage, to be able to include the claim queue in the storage proof used by the parachain. This way the parachain can report the `core` directly it wants to be scheduled on. We don't need any offset nor any selector.

I like this idea, I wanted at some point to do it like this, However there are 2 downsides that come to mind:
- PoV overhead from storing these proofs, even if they are small
- parachain needs to keep some state about what claims are used over a window of valid relay parents

But! there are some benefits:
- Without a qc offset, parachain blocks can be built ahead of time and can get in earlier
- Handles the case when availability stalls for a few blocks for a candidate on a core. Current proposal would not work anymore if the claim queue of the stalled core does not advance and claim queue of other cores assigned to the para are popped leading to different number of assignments at cq offset 0 that is used by runtime to compute the core index.
- The core index is fully deterministic, it doesn't depend on any relay chain variables anymore.

> 2. The UMP message is made optional. This way the runtime gives up the "control" to the PoV submitter to decide on the core. This gives us the freedom we have right now and will work for all chains that don't use elastic scaling without a problem.

 I don't really think we should give control to the PoV submitter to decide the core, this RFC is actually about removing that control for improved security, but I agree the RFC is a `no-op` for single core parachains. 
In any case if the parachain itself commits to a `CoreIndex` as discussed above you get all the freedom you have right now and for simplicity lets keep UMP commitments mandatory.

Reply:
A couple of things to do to fix in current proposal so that it fully works with availability timeouts:
- keep claim queue snapshots for all relay valid parents in the runtime
- check the core index using the snapshots rather than assuming cq offset 0

This would allow us to back things earlier and still determine a valid core index in the runtime if availability is slow.

Reply:
Actually a better idea is to pop the claim queue for the stalled cores and push the claim to the back of the queue: https://github.com/paritytech/polkadot-sdk/issues/4776#issuecomment-2309949169.

Reply:
> I also think that we will have problems with on demand? Let's assume we are able to buy a core and assign it in the same block (I know there exists somewhere an issue for this). How should the collator be able to build a block on time? Maybe making the UMP optional could be a solution for this?

One easy fix is to duplicate the claim for a order if the claim queue is empty. The parachain gets two chances if it uses cq offset `0` and has enough time to do well with cq offset `1`.

Reply:
> > I also think that we will have problems with on demand? Let's assume we are able to buy a core and assign it in the same block (I know there exists somewhere an issue for this). How should the collator be able to build a block on time? Maybe making the UMP optional could be a solution for this?
> 
> One easy fix is to duplicate the claim for a order if the claim queue is empty. The parachain gets two chances if it uses cq offset `0` and has enough time to do well with cq offset `1`.

Yes, this is planned to be done while solving https://github.com/paritytech/polkadot-sdk/issues/4776 (@eskimor also mentioned this)

Reply:
There is another issue with on-demand as @skunert pointed out today: Assignments can come up after the fact.

E.g. at relay parent X, the claim queue looks like this:

```
[
[A,B,C],
[A,A,C],
[],
]
```

Then the next relay parent something get scheduled for on-demand and above might become:

```
[
[B,C],
[A,C],
[E,C],
]
```

Messing up the position 0 check for parachain `C`.  Not sure yet how to solve this best. The issue is not unique to on-demand, can also happen with incomplete bulk assignments.

Reply:
Keeping a claim queue snapshot at valid relayparents in the runtime allows us to compute the core index correctly.

Reply:
Snapshots are fine. Possible future optimization (if necessary): Accompany each entry in the claim queue with the block number it came into existence. Then by only knowing the block number of the relay parent, we can filter to only the relevant entries.

Reply:
> One easy fix is to duplicate the claim for a order if the claim queue is empty. The parachain gets two chances if it uses cq offset `0` and has enough time to do well with cq offset `1`.

And if the claim queue is not empty? They loose their claim?

Reply:
> > One easy fix is to duplicate the claim for a order if the claim queue is empty. The parachain gets two chances if it uses cq offset `0` and has enough time to do well with cq offset `1`.
> 
> And if the claim queue is not empty? They loose their claim?

If the parachain is using offset 1, no. They have enough time to build the candidate.
If the parachain is using offset 0, then it can only do synchronous backing if my understanding is right.

Reply:
> And if the claim queue is not empty? They loose their claim?

No. The claim would not even be lost without the duplication on empty queues, but if the queue was empty before your claim immediately goes to position 0, which means only synchronous backing. Position 0 is therefore kind of "special".

There is no such problem if the queue was not empty. If at least one item was present before, you start at position 1, which already provides capability for full blocks.